### PR TITLE
[MM-47990] Improved app cleanup on uninstall; List fixes

### PR DESCRIPTION
**Pull request by levb**
#### Summary
- Added the cleanup of OAuth2 tokens and other user data on app uninstall
- Fixed faulty use of pluginapi.KV.ListKeys
- Refactored some List methods
- Added `/apps debug store list|pollute` command
- Adjusted the uninstall test for the changes

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-47990

### CODE REVIEW

File: test/restapitest/uninstall.go
Comment (javaguirre) [inline test/restapitest/uninstall.go:103]: Why 599 and 1401?

Comment (levb) [reply test/restapitest/uninstall.go:103]: The "window" size for paging is 1000, so I wanted to create 2K, but unevenly, in case the creation order matters for `ListKeys`. Really, Maybe should be 1401 and then 599.

**Review by mickmister — APPROVED** [review]
LGTM

File: server/store/session.go
Comment (hanzei) [inline server/store/session.go:83]: Why is this not correct? (Besides the check ` len(keys) < ListKeysPerPage` being an issue)

File: server/store/subscriptions.go
Comment (hanzei) [reply server/store/subscriptions.go:105]: And log a debug message?

**Review by hanzei — CHANGES_REQUESTED** [review]

File: server/store/subscriptions.go
Comment (levb) [reply server/store/subscriptions.go:105]: @mickmister a weird coincidence? Basically, impossible. If so, the record would be harmless and best be left alone. @hanzei at the moment, logging appears a bit impractical, and this is not really a reachable condition. Inclined to leave as is.

File: server/proxy/uninstall.go
Comment (levb) [reply server/proxy/uninstall.go]: Moved into a `defer`-ed cleanup. We _should_ disable the app if there was a failure to clean up its data.

File: server/store/kv_debug_info.go
Comment (levb) [reply server/store/kv_debug_info.go:115]: The issue is, with filters it returns fewer keys on normal results, stopping prematurely.

File: test/restapitest/uninstall.go
Comment (levb) [reply test/restapitest/uninstall.go:149]: No

File: server/store/session.go
Comment (levb) [reply server/store/session.go]: Restored

File: server/store/session.go
Comment (levb) [reply server/store/session.go:83]: I suppose it can come back if the check is done right, but why? I see little value in the wrapper, TBH.

**Comment by javaguirre** [comment]
@levb What's the most effective and easiest way of QA these changes?

**Comment by levb** [comment]
> @levb What's the most effective and easiest way of QA these changes?

@javaguirre Check the `restapitest` changes, maybe add more?

**Comment by javaguirre** [comment]
The commands seem to work fine, I'm checking the tests now to see if we might be missing something.

Screenshot 2022-12-07 at 10 13 20

**Comment by javaguirre** [comment]
Didn't like where I was going there, sorry for the force-push, only removed my commits, so no hashes changed.

**Comment by levb** [comment]
@javaguirre please avoid force-pushing, it makes history difficult to understand. If you need to update with master, or other branches - merge, do not rebase.

File: server/store/subscriptions.go
Comment (hanzei) [reply server/store/subscriptions.go:105]: I'm 0/5. It this is a weird edge case that should never happen, a log message does help with debugging things, if it ever does happen.

File: server/store/kv_debug_info.go
Comment (hanzei) [reply server/store/kv_debug_info.go:115]: Good point. Does that mean the pattern in general is correct?

Comment (hanzei) [reply server/store/kv_debug_info.go:115]: Does that mean the pattern in general is correct?

Comment (hanzei) [reply server/store/kv_debug_info.go:115]: To which filters are you referring here? The call to `mm.KV.ListKeys` does not include any filters.

Comment (levb) [reply server/store/kv_debug_info.go:115]: I don't think so, not with filters. Checking for an empty set is actually not guaranteed with filters either. I think the entire filters interface is broken, am I wrong?

**Comment by javaguirre** [comment]
> @javaguirre please avoid force-pushing, it makes history difficult to understand. If you need to update with master, or other branches - merge, do not rebase.

I know! I just used it to remove my commits 👍

**Comment by levb** [comment]
@hanzei ping? This can not be merged without your approval of the review/feedback changes.

File: server/proxy/uninstall.go
Comment (hanzei) [inline server/proxy/uninstall.go:49]: Nitpick: Why does the error need to be included two times? In  `WithError(err)` and via `Errorf`?

File: server/proxy/uninstall.go
Comment (hanzei) [inline server/proxy/uninstall.go:51]: Dito

File: server/proxy/uninstall.go
Comment (levb) [reply server/proxy/uninstall.go:49]: to make JSON-less text readable (i.e. include the essentials). However, perhaps errors.Cause(err).Error() is a better one to include, leaving the full one in JSON?

File: server/store/kv_debug_info.go
Comment (hanzei) [reply server/store/kv_debug_info.go:115]: To summarize:
1. If `pluginapi.WithPrefix` is used, the returned list may not contain `ListKeysPerPage` items and hence checking ` len(keys) < ListKeysPerPage`  to test if all keys have been fetched is wrong.
2. If `ListKeysPerPage` is used without a filter, such a check is correct.
3. By checking for `len(keys) == 0` one, additional RPC request is made (which fetches an empty list)

Do you agree with these three items? If so, is it worth changing the code to not make that additional RPC request?

File: server/store/subscriptions.go
Comment (levb) [reply server/store/subscriptions.go:105]: There's no logger easily available, maybe after cc-logger refactoring would be easier. Inclined to defer/pass.

**Review by hanzei — APPROVED** [review]

File: server/store/kv_debug_info.go
Comment (hanzei) [reply server/store/kv_debug_info.go:115]: Your are right, the operation is already quite expensive. The additional RPC does not matter much.

**Comment by hanzei** [comment]
@javaguirre Are you fine with merging this PR?

**Comment by levb** [comment]
@hanzei I addressed your most recent comments (error formatting). Just waiting on @javaguirre removing the "QA Review"? cc @DHaussermann

**Comment by javaguirre** [comment]
yes, I'm fine merging it

**Pull request merged**
